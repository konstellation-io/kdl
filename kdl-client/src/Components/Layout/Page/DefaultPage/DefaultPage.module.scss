@import 'Styles/colors';
@import 'Styles/variables';
@import 'Styles/mixins';

$sidebar-width: 40 * $grid-unit;

.container {
  position: relative;
  display: flex;
  width: 100%;
  height: 100%;
  background-color: palette(base, 700);
}

.actions div {
  padding: 0;
  flex-direction: row;
}

.sidebar {
  width: $sidebar-width;
  flex-shrink: 0;
  display: flex;
  flex-direction: column;
  background-color: palette(base, 500);
  padding: 10 * $grid-unit 5 * $grid-unit 0 3 * $grid-unit;

  h1 {
    text-align: left;
  }

  .subtitle {
    margin-top: 3 * $grid-unit;
  }

  .bottom {
    margin-top: auto;
  }
}

.content {
  display: flex;
  flex-direction: column;
  height: 100%;
  width: 100%;
  max-width: 1440px - $sidebar-width;
  background-color: palette(base, 800);
  overflow: auto;
}

.children {
  flex-grow: 1;
}
