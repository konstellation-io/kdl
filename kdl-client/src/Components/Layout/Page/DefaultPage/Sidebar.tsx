import React, { useRef } from 'react';

import styles from './DefaultPage.module.scss';

export let bottomRef: React.MutableRefObject<HTMLDivElement | null>;
export let topRef: React.MutableRefObject<HTMLDivElement | null>;

type Props = {
  title: string;
  subtitle?: string;
};
function Sidebar({ title, subtitle }: Props) {
  bottomRef = useRef<HTMLDivElement>(null);
  topRef = useRef<HTMLDivElement>(null);

  return (
    <div className={styles.sidebar}>
      <h1>{title}</h1>
      {subtitle && <h3 className={styles.subtitle}>{subtitle}</h3>}
      <div className={styles.top} ref={topRef} />
      <div className={styles.bottom} ref={bottomRef} />
    </div>
  );
}

export default Sidebar;
