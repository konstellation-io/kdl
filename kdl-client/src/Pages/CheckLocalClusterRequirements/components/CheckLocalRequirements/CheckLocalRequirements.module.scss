@import 'Styles/variables';
@import 'Styles/mixins';
@import 'Styles/colors';

$circle-diameter: $grid-unit;
$circle-top: 11px;
$circle-left: 0px;

.container {
  margin-bottom: 6 * $grid-unit;
}

.checkList ul {
  @include font-small;
  color: font-color(lowlight);

  padding: 0;

  li {
    transition: color ease 0.6s;
  }
}

@mixin status-circle($color) {
  color: $color;

  &:after {
    background-color: $color;
  }

  &:before {
    background-color: $color;
  }
}

.circle {
  @include circle(transparent);

  content: '';
  position: absolute;
  left: 0;
  top: 11px;
}

.requirement {
  list-style: none;
  position: relative;
  padding: $grid-unit 0 $grid-unit 2 * $grid-unit;

  &:before {
    @extend .circle;
  }

  &:after {
    @extend .circle;
  }

  &.pending {
    @include status-circle(palette(highlight));

    &:before {
      animation: loading-anim 2s ease 0s infinite;
    }
  }

  &.fail:before,
  &.ok:before {
    animation: loading-anim 0s ease, loading-anim 1s ease;
  }
  &.fail {
    @include status-circle(palette(error));
  }
  &.ok {
    @include status-circle(palette(success));
  }
}

.errorBox {
  position: relative;
  margin: 1.5 * $grid-unit 0 0 4 * $grid-unit;

  &:before,
  &:after {
    content: '';
    position: absolute;
  }

  &:before {
    top: -24px;
    left: -28px;
    width: 14px;
    height: 43px;
    border: dashed 1px palette(error);
    border-right-color: transparent;
    border-top-color: transparent;
  }
  &:after {
    width: 10px;
    height: 10px;
    border: 1px solid palette(error);
    border-top-color: transparent;
    border-left-color: transparent;
    top: 13px;
    left: -21px;
    transform: rotate(-33deg) skew(25deg);
  }
}

@mixin set-anim-state($diameter, $offset, $opacity) {
  width: $diameter;
  height: $diameter;
  top: $circle-top - $offset;
  left: $circle-left - $offset;
  opacity: $opacity;
}

$circle-increase: 1.8 * $grid-unit;
$circle-diameter-end: $circle-diameter + $circle-increase;
@keyframes loading-anim {
  0% {
    @include set-anim-state($circle-diameter, 0px, 0.6);
  }
  60%,
  100% {
    @include set-anim-state($circle-diameter-end, $circle-increase / 2, 0);
  }
}
