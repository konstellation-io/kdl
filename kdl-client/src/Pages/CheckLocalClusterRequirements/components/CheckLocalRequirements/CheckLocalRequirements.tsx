import React, { useEffect, useState } from 'react';

import { CheckState } from 'Pages/CheckLocalClusterRequirements/CheckLocalClusterRequirements';
import { cloneDeep } from 'lodash';
import cx from 'classnames';
import { ipcRenderer } from 'electron';
import styles from './CheckLocalRequirements.module.scss';

enum CheckId {
  k8s = 'k8s',
  minikube = 'minikube',
  helm = 'helm',
}

type CheckRequirementResponse = [CheckId, boolean];

type Check = {
  id: CheckId;
  label: string;
  isOk: boolean | null;
};

const generateInitialCheck = (id: CheckId, label: string) => ({
  id,
  label,
  isOk: null,
});

const initialCheckStatus: Check[] = [
  generateInitialCheck(CheckId.k8s, 'Kubernetes installed'),
  generateInitialCheck(CheckId.minikube, 'Minikube running'),
  generateInitialCheck(CheckId.helm, 'Helm available'),
];

type Props = {
  setChecksState: (newCheckState: CheckState) => void;
};

function CheckLocalRequirements({ setChecksState }: Props) {
  const [checks, setChecks] = useState(initialCheckStatus);

  useEffect(() => {
    const onRequirementReply = (
      _: unknown,
      [requirement, isOk]: CheckRequirementResponse
    ) => {
      setChecks((actChecks) => {
        const newCheckStatus = cloneDeep(actChecks);
        const actCheck = newCheckStatus.find((e) => e.id === requirement);
        if (actCheck) actCheck.isOk = isOk;

        return newCheckStatus;
      });
    };

    ipcRenderer.on('checkRequirementReply', onRequirementReply);

    checks.forEach((check) => {
      ipcRenderer.send('runCheckRequirement', check.id);
    });

    return () => {
      ipcRenderer.removeListener('checkRequirementReply', onRequirementReply);
    };
    // We want to execute this on on component mount/unmount
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  useEffect(() => {
    switch (true) {
      case checks.some((c) => c):
        // FIXME: Uncomment next line and remove previous.
        // case checks.every((c) => c.isOk):
        setChecksState(CheckState.OK);
        break;
      case checks.some((c) => c.isOk === false):
        setChecksState(CheckState.ERROR);
        break;
    }
  }, [checks, setChecksState]);

  const requirements = checks.map((check) => (
    <li
      key={check.id}
      className={cx(styles.requirement, {
        [styles.pending]: check.isOk === null,
        [styles.ok]: check.isOk,
        [styles.fail]: check.isOk === false,
      })}
    >
      {check.label}
    </li>
  ));

  return (
    <div className={styles.container}>
      <div className={styles.checkList}>
        <ul>{requirements}</ul>
      </div>
    </div>
  );
}

export default CheckLocalRequirements;
