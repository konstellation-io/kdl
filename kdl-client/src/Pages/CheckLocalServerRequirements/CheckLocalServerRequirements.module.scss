@import 'Styles/variables';
@import 'Styles/mixins';
@import 'Styles/colors';

.container {
  padding: 8 * $grid-unit;
}

.title {
  @include font-title;
  color: font-color(light);

  margin-bottom: 2 * $grid-unit;
}

.description {
  @include font-small;
  color: font-color(regular);

  margin-bottom: 4 * $grid-unit;
}

.installButton {
  width: 18 * $grid-unit;
}

.cancelButton {
  width: 14 * $grid-unit;
}
