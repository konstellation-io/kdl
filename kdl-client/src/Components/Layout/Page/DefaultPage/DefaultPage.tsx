import React, { FC } from 'react';

import ActionsBar from 'Components/Layout/ActionsBar/ActionsBar';
import Sidebar from './Sidebar';
import styles from './DefaultPage.module.scss';

type Props = {
  title: string;
  subtitle?: string;
  children: JSX.Element;
  actions?: JSX.Element | JSX.Element[];
};
const DefaultPage: FC<Props> = ({ title, subtitle, actions, children }) => (
  <div className={styles.container}>
    <Sidebar title={title} subtitle={subtitle} />
    <div className={styles.content}>
      <div className={styles.children}>{children}</div>
      {actions && <ActionsBar className={styles.actions}>{actions}</ActionsBar>}
    </div>
  </div>
);

export default DefaultPage;
