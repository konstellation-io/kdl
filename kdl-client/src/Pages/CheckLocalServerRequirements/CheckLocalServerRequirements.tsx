import React, { useState } from 'react';

import { Button } from 'kwc';
import CheckLocalRequirements from './components/CheckLocalRequirements/CheckLocalRequirements';
import DefaultPage from 'Components/Layout/Page/DefaultPage/DefaultPage';
import ROUTE from 'Constants/routes';
import styles from './CheckLocalServerRequirements.module.scss';
import { useHistory } from 'react-router-dom';

export enum CheckState {
  PENDING,
  OK,
  ERROR,
}

function CheckLocalServerRequirements() {
  const [checksState, setChecksState] = useState(CheckState.PENDING);
  const history = useHistory();

  const actions = [
    <Button
      label="CANCEL"
      key="button-2"
      className={styles.cancelButton}
      onClick={() => history.goBack()}
    />,
    <Button
      label="INSTALL"
      key="button-1"
      disabled={checksState !== CheckState.OK}
      to={ROUTE.INSTALL_LOCAL_SERVER}
      className={styles.installButton}
      primary
    />,
  ];
  return (
    <DefaultPage
      title="Install local Server"
      subtitle="Installing a local server allows the user to create fully functional KDL projects. You can stop the server at any time to free up resources."
      actions={actions}
    >
      <div className={styles.container}>
        <p className={styles.title}>Checking requirements</p>
        <p className={styles.description}>
          Before installing the local Server in your machine, some checks must
          be done. Installation will be done within a Kubernetes environment and
          may require the installation of different tools.
        </p>
        <CheckLocalRequirements setChecksState={setChecksState} />
      </div>
    </DefaultPage>
  );
}

export default CheckLocalServerRequirements;
