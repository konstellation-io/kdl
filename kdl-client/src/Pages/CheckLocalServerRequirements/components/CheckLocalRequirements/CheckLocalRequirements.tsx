import React, { Fragment, useEffect, useState } from 'react';

import { CheckState } from 'Pages/CheckLocalServerRequirements/CheckLocalServerRequirements';
import ErrorBox from 'Components/ErrorBox/ErrorBox';
import { cloneDeep } from 'lodash';
import cx from 'classnames';
import { ipcRenderer } from 'electron';
import styles from './CheckLocalRequirements.module.scss';

enum CheckId {
  k8s = 'k8s',
  minikube = 'minikube',
  helm = 'helm',
}

const checkError = {
  k8s: {
    title: 'Check 1 error',
    message: 'Check 1 message',
    docUrl: 'google.es',
  },
  minikube: {
    title: 'Check 2 error',
    message: 'Check 2 message',
    docUrl: 'google.es',
  },
  helm: {
    title: 'Check 3 error',
    message: 'Check 3 message',
    docUrl: 'google.es',
  },
};

type CheckRequirementResponse = [CheckId, boolean];

type Check = {
  id: CheckId;
  label: string;
  isOk: boolean | null;
  open: boolean;
};

const generateInitialCheck = (id: CheckId, label: string) => ({
  id,
  label,
  isOk: null,
  open: false,
});

const initialCheckStatus: Check[] = [
  generateInitialCheck(CheckId.k8s, 'Kubernetes installed'),
  generateInitialCheck(CheckId.minikube, 'Minikube running'),
  generateInitialCheck(CheckId.helm, 'Helm available'),
];

type Props = {
  setChecksState: (newCheckState: CheckState) => void;
};

function CheckLocalRequirements({ setChecksState }: Props) {
  const [checks, setChecks] = useState(initialCheckStatus);

  function resetCheck(checkId: string) {
    const newChecks = cloneDeep(checks);
    const target = newChecks.find((c) => c.id === checkId);
    if (target) target.isOk = null;

    setChecks(newChecks);
  }

  function checkRequirement(checkId: string) {
    resetCheck(checkId);
    ipcRenderer.send('checkRequirement', checkId);
  }

  useEffect(() => {
    const onRequirementReply = (
      _: unknown,
      [requirement, isOk]: CheckRequirementResponse
    ) => {
      setChecks((actChecks) => {
        const newCheckStatus = cloneDeep(actChecks);
        const actCheck = newCheckStatus.find((e) => e.id === requirement);
        if (actCheck) actCheck.isOk = isOk;

        return newCheckStatus;
      });
    };

    ipcRenderer.on('checkRequirementReply', onRequirementReply);

    checks.forEach((check) => {
      checkRequirement(check.id);
    });

    return () => {
      ipcRenderer.removeListener('checkRequirementReply', onRequirementReply);
    };
    // We want to execute this on on component mount/unmount
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  useEffect(() => {
    if (checks.every((c) => c.isOk)) {
      setChecksState(CheckState.OK);
    } else if (checks.some((c) => c.isOk === false)) {
      setChecksState(CheckState.ERROR);
    }
  }, [checks, setChecksState]);

  function toggleCheckOpen(idx: number, isOpening: boolean) {
    const newChecks = cloneDeep(checks);

    // Close all checks but opened
    newChecks.forEach((check) => (check.open = false));
    if (isOpening) newChecks[idx].open = true;

    setChecks(newChecks);
  }

  const requirements = checks.map((check, idx) => (
    <Fragment key={check.id}>
      <li
        className={cx(styles.requirement, {
          [styles.pending]: check.isOk === null,
          [styles.ok]: check.isOk,
          [styles.fail]: check.isOk === false,
        })}
      >
        {check.label}
      </li>
      {check.isOk === false && (
        <div className={styles.errorBox}>
          <ErrorBox
            title={checkError[check.id].title}
            message={checkError[check.id].message}
            docUrl={checkError[check.id].docUrl}
            onChange={() => toggleCheckOpen(idx, !check.open)}
            openController={check.open}
            action={{
              label: 'RETRY',
              onClick: () => checkRequirement(check.id),
            }}
          />
        </div>
      )}
    </Fragment>
  ));

  return (
    <div className={styles.container}>
      <div className={styles.checkList}>
        <ul>{requirements}</ul>
      </div>
    </div>
  );
}

export default CheckLocalRequirements;
