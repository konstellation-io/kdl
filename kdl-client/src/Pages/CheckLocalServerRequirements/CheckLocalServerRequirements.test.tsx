import CheckLocalServerRequirements, {
  CheckState,
} from './CheckLocalServerRequirements';

import { Button } from 'kwc';
import CheckLocalRequirements from './components/CheckLocalRequirements/CheckLocalRequirements';
import DefaultPage from 'Components/Layout/Page/DefaultPage/DefaultPage';
import React from 'react';
import { shallow } from 'enzyme';

const mockGoBack = jest.fn();
jest.mock('react-router-dom', () => ({
  useHistory: jest.fn(() => ({
    goBack: mockGoBack,
  })),
}));

const component = shallow(<CheckLocalServerRequirements />);

describe('CheckLocalServerRequirements page', () => {
  test('Component match snapshot', () => {
    expect(component).toMatchSnapshot();
  });

  it('show right texts', () => {
    expect(component.contains('Checking requirements')).toBeTruthy();
  });

  test('Install button is disabled when checks are not ok', () => {
    expect(
      component.find(DefaultPage).dive().find(Button).last().props().disabled
    ).toBeTruthy();
  });

  it('handles check updates', () => {
    component
      .find(CheckLocalRequirements)
      .props()
      .setChecksState(CheckState.OK);

    expect(
      component.find(DefaultPage).dive().find(Button).last().props().disabled
    ).toBeFalsy();
  });

  test('back action', () => {
    component.find(DefaultPage).dive().find(Button).first().simulate('click');

    expect(mockGoBack).toHaveBeenCalledTimes(1);
  });
});
