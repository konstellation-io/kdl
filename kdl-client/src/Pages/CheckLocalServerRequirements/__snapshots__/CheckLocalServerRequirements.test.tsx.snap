// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckLocalServerRequirements page Component match snapshot 1`] = `
<DefaultPage
  actions={
    Array [
      <P
        className="cancelButton"
        label="CANCEL"
        onClick={[Function]}
      />,
      <P
        className="installButton"
        disabled={true}
        label="INSTALL"
        primary={true}
        to="/install-local-server"
      />,
    ]
  }
  subtitle="Konstellation Server is based on Kubernetes. Kubernetes can run in almost any platform. The local server installation is checking if everything is ready to start the installation."
  title="Install local Server"
>
  <div
    className="container"
  >
    <p
      className="title"
    >
      Checking requirements
    </p>
    <p
      className="description"
    >
      In hac habitasse platea dictumst. Vivamus adipiscing fermentum quam volutpat aliquam. Integer et elit eget elit facilisis tristique. Nam vel iaculis mauris. Sed ullamcorper tellus.
    </p>
    <CheckLocalRequirements
      setChecksState={[Function]}
    />
  </div>
</DefaultPage>
`;
